_posts/2016-02-02-increasing-login-capcity-by-using-node-js.md

```diff
@@ -0,0 +1,72 @@
+---
+layout:     post
+title:      Increasing Login Capacity by Using Node.js
+date:       2016-02-02 20:21:00
+summary:    How we went about transitioning key parts of our stack to Node.js to benefit from the asynchronous nature that it brings.
+author:     tom_lomas
+category:   Product
+tags:       product, lamp, deployment, node, node.js, nginx, uptime, availability
+---
+
+Historically speaking, the sidebar that you may have come to use on Sky Betting & Gaming has been completely powered by a LAMP stack. It has worked well over the years and allows our customers to both login and register, but to also view account history, like open bets and casino spins.
+
+However, as demand increases, especially at crucial times of the year, we need to ensure that our platform remains available and page times are responsive. Login, for example, is used by nearly everyone that comes to the website - whether that is logging in fresh on a desktop browser, opening any of the apps or moving between websites (single sign on) in the Sky Betting & Gaming portfolio.
+
+We regularly load test key customer flows through the sidebar to gauge our throughput and using this enables us to get a good idea of our capacity. From this we can investigate any bottlenecks and identify any areas of focus. It also allows us to benchmark proposed changes and determine if they are really worth it from a performance standpoint.
+
+## Logging in
+
+Each of the websites, apps or services that integrate with us are referred to as a 'consumer' and on the busiest day of a normal week, usually a Saturday, logins across the consumers can run into the hundreds **per second**. This can increase astronomically on large sporting days, such as the Grand National.
+
+Each consumer requires different information about the customer. For example, [Sky Bet](https://www.skybet.com) may want to know the customer's balance whereas [Super 6](https://super6.skysports.com) may only want to know their unique customer ID. The amount of information that they require determines how many calls we need to make to our backend API and datastores.
+
+Traditionally all of this was handled in PHP. However, with that it meant that each call to find information about the customer (XML over curl) would block the next until it was complete.
+
+To keep up with the growth of the business and the new products launching, we needed to come up with a solution. The solution was Node.js.
+
+## Utilising the asychronous nature of Node.js in login
+
+When a login is performed it can require up to 4 separate calls to an XML based API for customer information, or even further datastores such as Couchbase. Traditionally this would mean that the 4th call is forced to wait for the 3 others to complete, even if it didn't care about the result of the earlier calls. As most of these can be ran asynchronously without needing the response of another first, Node.js suites us perfectly.
```

Comment:
"XML-based", rather than "XML based"
"Fourth", rather than "4th"
"Three", rather than "3"
"executed asynchronously", rather than "ran asynchronously"